unison-share-api/src/Unison/Server/CodebaseServer.hs

```diff
@@ -146,16 +147,72 @@ data BaseUrl = BaseUrl
     urlPort :: Port
   }
 
-data BaseUrlPath = UI | Api
+data NameOrHashRef = Name [Text] | Hash Text
+
+data DefinitionReference
+  = TermReference NameOrHashRef
+  | TypeReference NameOrHashRef
+  | AbilityConstructorReference NameOrHashRef
+  | DataConstructorReference NameOrHashRef
+
+data BaseUrlPath
+  = UI
+      { namespace :: Maybe [Text],
+        definition :: Maybe DefinitionReference
+      }
+  | Api
 
 instance Show BaseUrl where
   show url = urlHost url <> ":" <> show (urlPort url) <> "/" <> (URI.encode . unpack . urlToken $ url)
 
 urlFor :: BaseUrlPath -> BaseUrl -> String
-urlFor path baseUrl =
-  case path of
-    UI -> show baseUrl <> "/ui"
+urlFor baseUrlPath baseUrl =
```

Comment:
In general I think I'd recommend building URI's using the `URI` record and passing all the fields, but you're doing a good job of escaping everything so this is also probably fine, seems like it'd be really easy to accidentally end up missing/adding a `/` though, maybe build the path as a `[Text]` and `intercalate '/'` once it's ready? 😄